docs/proposals/train_api_proposal.md

```diff
@@ -0,0 +1,174 @@
+
+**<h3>New Train API proposal:</h3>**
+
+```python
+from kubeflow.training import TrainingClient
+# Arguments related to the model provider including credentials
+model_args = modelProviderClass()
+# Arguments related to the dataset provider including credentials
+dataset_args = datasetProviderClass()
+# Arguments related to the trainer code
+parameters = {key:value pairs}
+trainingClient.train(
+   workers=1, 
+   num_procs_per_worker = 1,
+   resources_per_worker={"gpu": "2", "cpu":8, "memory": "16Gi"}, 
+   model_args, 
+   dataset_args, 
+   parameters
+)
+```
+
+Example: 
+
+```python
+@dataclass
+class HuggingFaceModelParams:
+    access_token = field()
+    model = field()
+
+@dataclass
+class S3DatasetParams:
+    dataset = field()
+    eval_dataset = field()
+    access_token = field()
+    region = field()
+
+@dataclass
+class HuggingFaceTrainParams:
+    learning_rate = field()
+    peft_config = field()
+    transformerClass = field()
+
+trainingClient.train(
+   workers=1, 
+   num_procs_per_worker = 1,
+   resources_per_worker={"gpu": "2", "cpu":8, "memory": "16Gi"}, 
+   HuggingFaceModelParams(model='hf://openchat/openchat_3.5', access_token = "hf_..." ),
+   S3DatasetParams(dataset= 's3://doc-example-bucket1/train_dataset', eval_dataset = "s3://doc-example-bucket1/eval_dataset", access_token = "s3 access token", region="us-west-2"), 
+   HuggingFaceTrainParams(learning_rate=0.1, transformerClass="Trainer", peft_config = {})
+)
+```
+
+The new proposed API takes following arguments 
+
+1. System parameters - Number of workers, number of resources per workers(GPUs per worker).
+2. Model parameters - Model provider and repository details.
+3. Dataset parameters - Dataset provider and dataset details.
+4. Training parameters - Training specific parameters like learning rate etc.
+
+These parameters will be passed as container args or environment variables.
+
+**<h3>Implementation</h3>**
+
+1. Setup **init** **containers** that download the model and dataset to a PVC. Based on the specified model provider, corresponding training utility functions will be used. Eg: For Huggingface provider, Huggingface trainer can be used. For this **get_pytorchjob_template** function in the sdk needs to be changed to add init containers spec.. Inorder to download models and data sets, we need to support different providers like kaggle, hugging face, s3 or git lfs. The data can be stored in a shared volume between the init container and the main container.<br /> <br /> This way to download models allows using ReadWriteOnce and ReadOnlyMany PVCs. If we adopt the way of creating batch/v1 Job to download models to PVC, we need to force users to prepare ReadWriteOnce and  ReadOnlyMany PVCs.<br /> <br /> A new folder containing the code for downloading model and dataset can be added to generate the images for init_containers. Abstract classes will be used as base to create when dataset download, model download and training loop is written for base images.
```

Comment:
I think, we didn't mention how we are going to bypass user settings to the initContainer for storage and model initialization. 
Are we going to pass it via env variable or init container arguments ?

Reply:
done